opencv: add new version, variant, and patch

glennpj opened the pull request:
- added version 4.5.4
- added tesseract variant
- added patch to not add system paths

glennpj committed:
opencv: add new version, variant, and patch

- added version 4.5.4
- added tesseract variant
- added patch to not add system paths

adamjstewart commented on var/spack/repos/builtin/packages/opencv/cmake_no-system-paths.patch:
```diff
@@ -0,0 +1,59 @@
+-    endif()
+-  endif()
+-endif()
++#if(UNIX AND NOT ANDROID)
```
I would just delete it instead of commenting it out, smaller patch

adamjstewart commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -365,6 +371,15 @@ def cmake_args(self):
 
+        if '+tesseract' in spec:
+            tesseract = spec['tesseract']
+            leptonica = spec['leptonica']
```
If `leptonica` is a directly dependency (we link to it), we should have an explicit dependency on it instead of relying on the fact that `tesseract` currently has a dependency on it.

glennpj replied:
This is needed by the `FindTesseract.cmake` file. There is a related comment in that file:
```
# Tesseract OCR
ocv_check_modules(Tesseract tesseract) # lept is excluded (not a direct dependency)
```
However, Tesseract will not be found without finding leptonica, which is a dependency of tesseract. I am not sure that leptonica will ever _not_ be a dependency of tesseract but I suppose a replacement could surface. I think it makes sense to add it explicitly as it is tested for explicitly.

adamjstewart reviewed (commented):
Tesseract doesn't seem to be in OpenCV proper like all other components/modules we have variants for, it seems to be in opencv_contrib. Since is the first time we've added a opencv_contrib-specific component, I'll have to think if this is the best way to do this. We'll definitely need a conflict for `+tesseract` when `~contrib`.

glennpj commented:
> We'll definitely need a conflict for `+tesseract` when `~contrib`.

Yes, sorry, I completely missed that.

glennpj committed:
Add leptonica depends and contrib conflicts

adamjstewart commented:
> Tesseract doesn't seem to be in OpenCV proper like all other components/modules we have variants for, it seems to be in opencv_contrib. Since is the first time we've added a opencv_contrib-specific component, I'll have to think if this is the best way to do this.

@bvanessen do you have any thoughts on this? I'm thinking we might just want to make it a separate variant (not in modules or components).

glennpj commented:
I just discovered that there seems to be a problem with the installation of opencv that, as far as I can tell, is independent of this PR. Looking at the installation directory, the programs are not getting installed in `${prefix}/bin`. This is true for an installation of 4.5.2 that I did back in July as well.

glennpj commented:
Actually, I think things are fine. I did not realize that literally every variant is turned off.

glennpj commented:
I converted this to "draft" status as I found some components that do not build due to changed variable names. It looks like this began with version 4.5.2 but it seems like fixing those here is appropriate.

glennpj committed:
Add dependencies for 1394 support

- new package: libraw1394
- add sdl dependency to libdc1394
- add conflict for openjpeg and jasper

glennpj committed:
Adjust dependencies and conflicts for opencv modules

glennpj committed:
rewrite of opencv

- all prebuilt apps are now variants and can be installed
- core is no longer a variant. It was always built anyway so it was not
  really a variant.
- contrib is no longer a variant. All of the contrib modules are now
  available as variants.
- components that can not be built with Spack are no longer variants.
  They are set to 'off' to prevent pulling from system.
- handle the case where a module and a component have the same name
- use `with when` framework
- adjust dependencies and conflicts
- new package: libraw1394
- have libdc1394 depend on libraw1394
- patch to find clp
- patch to find onnx
- patch for cvv to find Qt
- format with black

glennpj commented:
The changes wound being significant but applications and contrib modules are now available as variants, while components that can not be built are no longer variants.

Summary:

    - all prebuilt apps are now variants and can be installed
    - core is no longer a variant. It was always built anyway so it was not
      really a variant.
    - contrib is no longer a variant. All of the contrib modules are now
      available as variants.
    - components that can not be built with Spack are no longer variants.
      They are set to 'off' to prevent pulling from system.
    - handle the case where a module and a component have the same name
    - use `with when` framework
    - adjust dependencies and conflicts
    - new package: libraw1394
    - have libdc1394 depend on libraw1394
    - patch to find clp
    - patch to find onnx
    - patch for cvv to find Qt
    - format with black

adamjstewart commented on var/spack/repos/builtin/packages/libdc1394/package.py:
```diff
@@ -10,17 +10,30 @@ class Libdc1394(AutotoolsPackage):
         args = []
-        args.append('--disable-dependency-tracking')
-        args.append('--disable-examples')
-        args.append('--disable-sdltest')
+        args.extend(
+            [
+                "--disable-dependency-tracking",
+                "--disable-examples",
+                "--disable-sdltest",
+            ]
+        )
         return args
```
You could simplify this by saying:
```python
return [
    ...
]
```

glennpj replied:
Good to know. I always assumed that it was standard format to have it explicit.

adamjstewart commented on var/spack/repos/builtin/packages/libdc1394/package.py:
```diff
@@ -10,17 +10,30 @@ class Libdc1394(AutotoolsPackage):
+    url = "https://downloads.sourceforge.net/project/libdc1394/libdc1394-2/2.2.6/libdc1394-2.2.6.tar.gz"
 
-    maintainers = ['traversaro']
+    maintainers = ["traversaro"]
```
@traversaro can you review these dependency changes?

traversaro replied:
Ok! Sorry for the late reply.

adamjstewart commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -2,399 +2,1105 @@
+    url = "https://github.com/opencv/opencv/archive/4.5.0.tar.gz"
+    git = "https://github.com/opencv/opencv.git"
+
+    maintainers = ["bvanessen", "adamjstewart"]
```
@glennpj want to join the party?

glennpj replied:
I ended up doing so.

adamjstewart commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -2,399 +2,1105 @@
+    version(
+        "3.2.0",
+        sha256="9541efbf68f298f45914b4e837490647f4d5e472b4c0c04414a787d116a702b2",
+        deprecated=True,
```
You can actually remove these deprecated versions now if you want

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -2,399 +2,1105 @@
+    patch("dnn_cuda.patch", when="@3.3.0:3.4.1+cuda+dnn")
+
+    patch("opencv3.2_cmake.patch", when="@3.2:3.4.1")
+    patch("opencv3.2_compiler_cmake.patch", when="@3.2")
+    patch("opencv3.2_vtk.patch", when="@3.2+vtk")
+    patch("opencv3.2_regacyvtk.patch", when="@3.2+vtk")
+    patch("opencv3.2_ffmpeg.patch", when="@3.2+videoio")
+    patch("opencv3.2_python3.7.patch", when="@3.2+python3")
+    patch("opencv3.2_fj.patch", when="@3.2 %fj")
```
If you drop 3.2 and older, a lot of these patches can be removed

glennpj replied:
Done.

adamjstewart commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -2,399 +2,1105 @@
     # Defined in `modules/*/CMakeLists.txt` using
     # `ocv_add_module(...)` and `ocv_define_module(...)`
     modules = [
-        'apps', 'calib3d', 'core', 'dnn', 'features2d', 'flann', 'gapi', 'highgui',
```
If you're going to remove variants, you'll need to update packages that depend on them. For example:
```console
$ grep depends_on.*opencv.*core */package.py
3dtk/package.py:    depends_on('opencv+calib3d+contrib+core+features2d+highgui+imgcodecs+imgproc+ml+videoio', when='+opencv')
caffe/package.py:    depends_on('opencv@3.2.0:3.4.12+core+highgui+imgproc+imgcodecs', when='+opencv')
candle-benchmarks/package.py:    depends_on('opencv@3.2.0: +core +highgui +imgproc +jpeg +png +tiff -dnn ~eigen ~gtk')
dbow2/package.py:    depends_on('opencv+calib3d+core+features2d+highgui+imgproc')
dlib/package.py:    depends_on('opencv+calib3d+core+features2d+highgui+imgproc+imgcodecs')
lbann/package.py:    depends_on('opencv@4.1.0: build_type=RelWithDebInfo +core +highgui '
mivisionx/package.py:    depends_on('opencv@:3.4 +calib3d+core+features2d+highgui+imgcodecs+imgproc+video+videoio+flann+photo+objdetect', type='build')
mxnet/package.py:    depends_on('opencv+core+highgui+imgproc+imgcodecs', when='+opencv')
openvslam/package.py:    depends_on('opencv@3.3.1:+core+imgcodecs+videoio+features2d+calib3d+highgui')
```

glennpj replied:
Yes, that is important. Missed the forest for the trees here. I added a new commit to address this, but the `cntk` package will need more work to get it to build.

adamjstewart commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -2,399 +2,1105 @@
-    depends_on('py-numpy', when='+python3', type=('build', 'run'))
-    depends_on('ffmpeg', when='+videoio')
-    depends_on('mpi', when='+videoio')
+    depends_on("gmake", type="build")
```
Is GNU Make actually required, or can it be any Make? Unless there's a GNUmakefile somewhere, I would remove this dep.

glennpj replied:
I guess probably not. The output said that it found `gmake`, which I interpreted to mean that it was looking specifically for `gmake`. On my system, `gmake` is a symlink to `make` and I did not test with `bmake`. I removed the dependency.

glennpj committed:
Incorporate recommended changes

- fix variants and dependencies on packages that depend on opencv
- remove opencv-3.2 and patches
- add some new patches to handle different versions
- cntk needs further work
- the openvslam package was markde deprecated as it is no longer an
  active project and the repository has no code

glennpj committed:
Remove gmake dependency.

traversaro commented on var/spack/repos/builtin/packages/libdc1394/package.py:
```diff
@@ -10,17 +10,26 @@ class Libdc1394(AutotoolsPackage):
-    depends_on('libusb')
+    depends_on("pkgconfig", type="build")
+    depends_on("libusb")
+    depends_on("sdl2")
```
If I am not wrong, `sdl`  is just a dependency for the examples, that are disabled in this recipe, no?

glennpj replied:
@traversaro Looking at the library links, I believe you are correct. I will remove that dependency.

glennpj committed:
Remove sdl support

SDL is only used in an example case, but the examples are not built.

adamjstewart commented:
Looks like there are some conflicts, can you resolve those?

traversaro reviewed (commented):
The `libdc1394` changes look ok for me @adamjstewart .

glennpj commented:
> Looks like there are some conflicts, can you resolve those?

It looks like there were permissions changes on several files in the `develop` branch that were put in place after my last update of `develop` that this PR was created from. A rebase seems to have resolved it.

adamjstewart commented on var/spack/repos/builtin/packages/3dtk/package.py:
```diff
@@ -37,7 +37,7 @@ class _3dtk(CMakePackage):
     depends_on('cmake@2.6.1:2', when='@1.2', type='build')
     depends_on('ninja', type='build')
-    depends_on('boost+serialization+graph+regex+filesystem+system+thread+date_time+program_options')
+    depends_on('boost@:1.75+serialization+graph+regex+filesystem+system+thread+date_time+program_options')
```
This change doesn't seem related to this PR, I would remove it

glennpj replied:
There is another change to this package that is related to this PR. If I split out that one change into another PR then either that PR or this PR will need to be rebased. It seems reasonable to handle this change here as this change is necessary to build the package.

glennpj replied:
I went ahead and split it, #27806.

adamjstewart commented on var/spack/repos/builtin/packages/bohrium/package.py:
```diff
@@ -68,6 +68,8 @@ class Bohrium(CMakePackage, CudaPackage):
     conflicts('~node~proxy')
     conflicts('~openmp~opencl~cuda')
 
+    conflicts('+cbridge', when='~python')
```
Same here, let's try to keep this PR focused only on OpenCV changes. Any other improvements can be made and reviewed in a separate PR

glennpj replied:
As above, there are changes to the bohrium package that are directly related to this PR. Other changes are necessary to get bohrium to build so splitting the changes will force PR dependencies and rebasing.

glennpj replied:
Split bohrium changes, depends on #27807.

adamjstewart commented on var/spack/repos/builtin/packages/dlib/package.py:
```diff
@@ -15,7 +15,4 @@ class Dlib(CMakePackage):
-    # Because concretizer is broken...
-    # TODO: remove when original concretizer is obsolete
-    depends_on('opencv+flann')
+    depends_on('opencv+calib3d+features2d+flann+highgui+imgproc+imgcodecs')
```
I think you can remove `+flann` from here, the new concretizer is the default now

adamjstewart commented on var/spack/repos/builtin/packages/openvslam/package.py:
```diff
@@ -10,15 +10,17 @@ class Openvslam(CMakePackage):
 
-    version('master', branch='master')
+    # This project consists only of a README file, that announces the
+    # termination of the project.
```
Sad, I was the one who added this package for a class I was taking. I don't think I was ever able to get it to build anyway, so I'm fine with removing it.

adamjstewart commented on var/spack/repos/builtin/packages/py-astropy/package.py:
```diff
@@ -62,7 +62,7 @@ class PyAstropy(PythonPackage):
     depends_on('erfa')
     depends_on('wcslib')
-    depends_on('cfitsio')
+    depends_on('cfitsio@:3')
```
Unrelated to this PR

glennpj replied:
This is related in the sense that the above is needed to build. However, it is a second level dependency so split out to #27805.

adamjstewart commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -2,399 +2,1062 @@
+
+    # components and modules with the same name
+    # used in `def cmake_args(self)`
+    component_and_module = ["freetype", "julia", "matlab"]
```
Can you clarify this? For these components/modules, is it possible to enable one and disable the other in CMake? If so, we should find a way to do this in the Spack package. If we have to, we can convert components/modules/etc to multi-valued variants to avoid naming conflict.

glennpj replied:
I structured this such that the variant is for the opencv module, and the supporting component, or external library, is set based on that. For instance, if the freetype variant is selected that does the following:
- sets the freetype opencv module to be built
- sets a dependency on freetype
- ensures that the cmake flag to use the spack freetype package is set

It seems reasonable to tie those together, as we want to use the spack freetype package if the opencv freetype module is built. Likewise, there would be no point in including a spack freetype package if the opencv freetype module is not built. I do not see how multi-valued variants would solve this, and in fact, I believe would make it more confusing for someone installing opencv and looking at the wall of variants. My goal was to present only one freetype variant, not two.

adamjstewart reviewed (commented):
Can you remove any changes that aren't directly related to OpenCV? This PR is hard enough to review as it is without including additional build fixes. These things can be moved to separate PRs.

glennpj committed:
remove openvslam

glennpj committed:
Remove opencv+flann variant from 3dtk

glennpj committed:
Back out cfitsio constraint from py-astropy

glennpj committed:
remove opencv+flann variant from dlib

glennpj committed:
remove boost constraint from 3dtk

glennpj committed:
Remove non-opencv related bohrium changes

glennpj committed:
Adjustments for cntk

- protobuf constraint at version 3.10
- need specific variants for opencv
- improve patch

glennpj committed:
Deprecate CNTK package

glennpj committed:
variant tweaks

- added appropriate conflicts for cublas
- made cuda/cudev relationship explicit
- moved openx to pending components as it needs an openvx package

glennpj committed:
fix isort style error

glennpj committed:
Use date version from kaldi rather than commit

adamjstewart commented on var/spack/repos/builtin/packages/3dtk/package.py:
```diff
@@ -37,7 +37,7 @@ class _3dtk(CMakePackage):
     depends_on('cmake@2.6.1:2', when='@1.2', type='build')
     depends_on('ninja', type='build')
-    depends_on('boost@:1.75+serialization+graph+regex+filesystem+system+thread+date_time+program_options')
+    depends_on('boost+serialization+graph+regex+filesystem+system+thread+date_time+program_options')
```
Is this undoing #27806?

adamjstewart commented on var/spack/repos/builtin/packages/bohrium/package.py:
```diff
@@ -86,24 +84,18 @@ class Bohrium(CMakePackage, CudaPackage):
     depends_on('py-cython', type="build", when="+python")
-    depends_on('py-virtualenv', type="build", when="+python")
-    depends_on('py-pip', type="build", when="+python")
-    depends_on('py-wheel', type="build", when="+python")
```
How are these changes related to this PR?

adamjstewart commented on var/spack/repos/builtin/packages/cntk/package.py:
```diff
@@ -49,6 +54,26 @@ class Cntk(Package):
+    # beyond 4.8.5.
+    conflicts('%gcc@5:')
+
+    def patch(self):
```
These changes don't seem related to this PR. If you found and fixed a build issue with `cntk`, that's great, but it could go in a different PR. This PR should only include changes to OpenCV and changes to other packages that result from that.

adamjstewart commented on var/spack/repos/builtin/packages/libdc1394/package.py:
```diff
@@ -10,17 +10,22 @@ class Libdc1394(AutotoolsPackage):
-    depends_on('libusb')
+    depends_on("pkgconfig", type="build")
+    depends_on("libusb")
+    depends_on("libraw1394")
```
Changes to this package should go in a different PR

adamjstewart commented on var/spack/repos/builtin/packages/libraw1394/package.py:
```diff
@@ -0,0 +1,19 @@
+from spack import *
+
+
+class Libraw1394(AutotoolsPackage):
```
Should be in a different PR

adamjstewart commented on var/spack/repos/builtin/packages/openvslam/package.py:
```diff
@@ -1,33 +0,0 @@
-# SPDX-License-Identifier: (Apache-2.0 OR MIT)
-
-
-class Openvslam(CMakePackage):
```
This package should be deprecated before being removed in case someone is still using it

glennpj replied:
See https://github.com/spack/spack/pull/27374#discussion_r759480940

adamjstewart replied:
By "remove" I mean deprecate first and delete in a later Spack release

glennpj replied:
Even though there is no code, just a README?

adamjstewart commented on var/spack/repos/builtin/packages/py-astropy/package.py:
```diff
@@ -62,7 +62,7 @@ class PyAstropy(PythonPackage):
     depends_on('erfa')
     depends_on('wcslib')
-    depends_on('cfitsio@:3')
+    depends_on('cfitsio')
```
Unrelated to this PR (reverting a different PR?)

adamjstewart commented on var/spack/repos/builtin/packages/py-mmcv/package.py:
```diff
@@ -23,5 +23,7 @@ class PyMmcv(PythonPackage):
     depends_on('opencv+python3', type=('build', 'run'))
     depends_on('py-cython', type='build')
+    depends_on('py-pip', type='build')
+    depends_on('py-wheel', type='build')
```
Separate PR

adamjstewart commented on var/spack/repos/builtin/packages/py-torchgeo/package.py:
```diff
@@ -56,10 +56,7 @@ class PyTorchgeo(PythonPackage):
-        depends_on('py-laspy@2:', when='@0.2:', type='run')
-        depends_on('open3d@0.11.2:+python', when='@0.2:', type='run')
-        depends_on('opencv+python3+imgcodecs+tiff+jpeg+png', type='run')
-        depends_on('py-pandas@0.19.1:', when='@0.2:', type='run')
```
I develop this library and I can assure you that these other deps are still needed lol

adamjstewart commented on var/spack/repos/builtin/packages/saga-gis/package.py:
```diff
@@ -68,10 +68,9 @@ class SagaGis(AutotoolsPackage, SourceforgePackage):
+    depends_on('opencv+jpeg+video+objdetect+ml+openmp', when='+opencv')
     # Set jpeg provider (similar to #8133)
-    depends_on('libjpeg', when='+opencv')
+    depends_on('jpeg', when='+opencv')
```
Separate PR

adamjstewart reviewed (commented):
This PR should contain:

1. changes to the OpenCV package
2. changes to other packages that are required as a result (i.e. removing `+core`)

This PR should not contain changes to other packages that are not related to OpenCV (i.e. fixing a broken build). This makes it harder for me to review. Also, it looks like there was a bad rebase, so this PR is actually undoing many of the changes that have been made in other PRs.

glennpj commented:
> This PR should contain:
> 
> 1. changes to the OpenCV package
> 2. changes to other packages that are required as a result (i.e. removing `+core`)
> 
> This PR should not contain changes to other packages that are not related to OpenCV (i.e. fixing a broken build). This makes it harder for me to review. Also, it looks like there was a bad rebase, so this PR is actually undoing many of the changes that have been made in other PRs.

I moved quite a few packages out of this PR into other PRs earlier, as you had requested. I agree that something seemed to have gone horribly wrong with the rebase and I wound undoing much of that. I will have to sort that out, sorry for the extra work.

glennpj commented:
The problem was not the rebase per se, but a git mistake I made when I moved packages to separate PRs. After creating the other PRs I added commits to this PR to remove the packages. When I rebased, those commits replayed on top and effectively put the packages back in this PR, **and** reversed the desired commits from the other PRs. I suppose one could argue that doing a rebase was a mistake as I could have merged in the develop branch and resolved conflicts from there. Either way, I did not manage git properly and I apologize for the confusion and extra work.

glennpj committed:
Revert changes from a bad rebase

glennpj commented:
- [ ] restore to "good" state
- [ ] create PR for 3dtk
- [ ] create PR for bohrium
- [ ] create PR for caffe

glennpj commented:
The `3dtk` package needs the new `aruco` variant in this PR.

glennpj commented:
The `bohrium` package needs the new `cudev` variant.

glennpj commented:
The `py-torchgeo` package needs the new `datasets` variant.

glennpj committed:
Add +flann to 3dtk and dlib

glennpj committed:
Merge branch 'develop' into opencv

adamjstewart commented on var/spack/repos/builtin/packages/opencv/package.py:
```diff
@@ -2,399 +2,1069 @@
+    conflicts("+v4l", when="platform=darwin", msg="Linux only")
+    conflicts("+win32ui", when="platform=darwin", msg="Windows only")
+    conflicts("+win32ui", when="platform=linux", msg="Windows only")
+    conflicts("+win32ui", when="platform=cray", msg="Windows only")
```
We can now make the variant conditional on the OS, although that might be difficult to do with the for-loop.

adamjstewart reviewed (commented):
Testing the build and some packages I care about now

adamjstewart commented on var/spack/repos/builtin/packages/py-torchgeo/package.py:
```diff
@@ -58,7 +58,7 @@ class PyTorchgeo(PythonPackage):
         depends_on('py-laspy@2:', when='@0.2:', type='run')
         depends_on('open3d@0.11.2:+python', when='@0.2:', type='run')
-        depends_on('opencv+python3+imgcodecs+tiff+jpeg+png', type='run')
+        depends_on('opencv+python3+imgcodecs+tiff+jpeg+png+datasets', type='run')
```
Tried testing this. The only dataset in TorchGeo that uses OpenCV is the LandCover.ai dataset. It uses this to read TIFF files and write JPEG/PNG files, all in pure-Python. However, this seems to fail:
```console
$ spack install py-torchgeo+datasets
$ spack load py-torchgeo+datasets
$ python
>>> from torchgeo.datasets import LandCoverAI
>>> ds = LandCoverAI(download=True, checksum=True)
Downloading https://landcover.ai/download/landcover.ai.v1.zip to data/landcover.ai.v1.zip
1538212864it [05:22, 4766024.88it/s]                                                                                                                                      
Extracting data/landcover.ai.v1.zip to data
[ WARN:0] global /var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/Adam/spack-stage/spack-stage-opencv-4.5.4-urkmryldyvjy5qmhjepejt6utj7qmnmg/spack-src/modules/imgcodecs/src/grfmt_tiff.cpp (457) readData OpenCV TIFF: TIFFRGBAImageOK: Sorry, requested compression method is not configured
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/Adam/torchgeo/torchgeo/datasets/landcoverai.py", line 109, in __init__
    self._verify()
  File "/Users/Adam/torchgeo/torchgeo/datasets/landcoverai.py", line 200, in _verify
    self._extract()
  File "/Users/Adam/torchgeo/torchgeo/datasets/landcoverai.py", line 234, in _extract
    exec(split)
  File "<string>", line 27, in <module>
AttributeError: 'NoneType' object has no attribute 'shape'
```
Is there something else we're missing here? Also, what made you add `+datasets`?

glennpj replied:
I guess I made an assumption about `datasets`. That warning message is coming from `libtiff`. Not sure if that leads to the error or not.

adamjstewart replied:
@glennpj this is a bit off-topic, but do you know how to determine which `libtiff` variants are needed? All I can tell is that it seems to use JPEG:
```console
$ file M-33-20-D-d-3-3.tif
M-33-20-D-d-3-3.tif: TIFF image data, little-endian
$ gdalinfo M-33-20-D-d-3-3.tif
...
Image Structure Metadata:
  COMPRESSION=YCbCr JPEG
  INTERLEAVE=PIXEL
  SOURCE_COLOR_SPACE=YCbCr
```
I've tried both `+jpeg` and `+old-jpeg` but I still can't read this file.

glennpj replied:
All of the variants for `libtiff` are set to `False`. That is surprising, but then opencv is the same way. For libtiff, compression support will need to be enabled. I did not realize libtiff was set that way so opencv will need to depend on variants being set when using tiff.

glennpj replied:
At a minimum, `libdeflate` is needed. I think `lzma` and `zlib` are going to be very common as well.

glennpj replied:
Using `libtiff+jpeg+libdeflate`
```
$ python
Python 3.9.9 (main, Jan 20 2022, 15:07:33)
[GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from torchgeo.datasets import LandCoverAI
>>> ds = LandCoverAI(download=True, checksum=True)
Downloading https://landcover.ai/download/landcover.ai.v1.zip to data/landcover.ai.v1.zip
1538212864it [01:20, 19074956.52it/s]
Extracting data/landcover.ai.v1.zip to data
Processed M-33-20-D-c-4-2 1/41
Processed M-33-20-D-d-3-3 2/41
Processed M-33-32-B-b-4-4 3/41
Processed M-33-48-A-c-4-4 4/41
Processed M-33-7-A-d-2-3 5/41
Processed M-33-7-A-d-3-2 6/41
Processed M-34-32-B-a-4-3 7/41
Processed M-34-32-B-b-1-3 8/41
Processed M-34-5-D-d-4-2 9/41
Processed M-34-51-C-b-2-1 10/41
Processed M-34-51-C-d-4-1 11/41
Processed M-34-55-B-b-4-1 12/41
Processed M-34-56-A-b-1-4 13/41
Processed M-34-6-A-d-2-2 14/41
Processed M-34-65-D-a-4-4 15/41
Processed M-34-65-D-c-4-2 16/41
Processed M-34-65-D-d-4-1 17/41
Processed M-34-68-B-a-1-3 18/41
Processed M-34-77-B-c-2-3 19/41
Processed N-33-104-A-c-1-1 20/41
Processed N-33-119-C-c-3-3 21/41
Processed N-33-130-A-d-3-3 22/41
Processed N-33-130-A-d-4-4 23/41
Processed N-33-139-C-d-2-2 24/41
Processed N-33-139-C-d-2-4 25/41
Processed N-33-139-D-c-1-3 26/41
Processed N-33-60-D-c-4-2 27/41
Processed N-33-60-D-d-1-2 28/41
Processed N-33-96-D-d-1-1 29/41
Processed N-34-106-A-b-3-4 30/41
Processed N-34-106-A-c-1-3 31/41
Processed N-34-140-A-b-3-2 32/41
Processed N-34-140-A-b-4-2 33/41
Processed N-34-140-A-d-3-4 34/41
Processed N-34-140-A-d-4-2 35/41
Processed N-34-61-B-a-1-1 36/41
Processed N-34-66-C-c-4-3 37/41
Processed N-34-77-A-b-1-4 38/41
Processed N-34-94-A-b-2-4 39/41
Processed N-34-97-C-b-1-2 40/41
Processed N-34-97-D-c-2-4 41/41
>>>
```

adamjstewart replied:
It looks like `+jpeg+zlib` is required. I'll add that in a separate PR. Anyway, you can remove `+datasets` from this PR, it isn't needed.

adamjstewart replied:
Hmm, I got it working with `+jpeg+zlib~libdeflate`, not sure if `+libdeflate` is actually required.

Yeah, the variants for `libtiff` are pretty bad. I'm the one who disabled everything by default in both OpenCV and libtiff. Trying to minimize building time for packages that only need a small portion of the library. Once we have public binaries we can expand this.

glennpj replied:
Right, zlib is needed for libdeflate, and clingo enables zlib support with libdeflate.

glennpj committed:
Use compression support with libtiff

glennpj committed:
remove `+datasets` from opencv dependency 

The py-torchgeo package does not need opencv+datasets.

glennpj committed:
fix typo

zip --> zlib

adamjstewart reviewed (approved):
I haven't looked through the changes in extreme detail (like the conflicts stuff) but I tested that the package installs and works with the only project I use it for (TorchGeo). I'm fine with merging this in its current state unless there are any remaining tests or reviewers you would like to wait for.

glennpj commented:
I have not put openvslam back in yet. I am not really sure what it means to have a package that has no code, even deprecated packages should be buildable, I think. Otherwise, I think this is ready to go.

adamjstewart commented:
Yeah, it's mostly for people who already have the package in their source cache. But I'm not even sure we use the source cache when downloading from `master` anway.

glennpj commented:
This is a bit off topic but I looked into openvslam a little more. The only version in spack is `master` and that is when I discovered that there was only a README. However, there are releases that could be pulled. Interestingly, the README has been updated recently with a link to a lengthy discussion about legal issues and academic misconduct. I did not read the whole issue but it makes me think more firmly that we should remove it from spack.

https://github.com/OpenVSLAM-Community/openvslam/blob/in-discussion/README.md

So, I say let's merge this PR. We can deal with openvslam later if desired.

svenevs commented:
Nice work @glennpj :slightly_smiling_face:  I'm not in a position to update my `spack` right now so I just bludgeoned in `sfm` after realizing I didn't get it with my install a few months ago.  Took a quick look to see if current `develop` had anything related to this module, you totally nailed it!  This package is looking really great now :heart:  Looking forward to the next time I get to re-tool things.
